spring-cloud-commons/src/test/java/org/springframework/cloud/client/hedger/HedgedRequestsExchangeFilterFunctionTests.java

```diff
@@ -0,0 +1,215 @@
+			Mono<ClientResponse> second = Mono.never();
+			when(exchangeFunction.exchange(clientRequest)).then(invocation -> count.getAndIncrement() == 0 ? first : second);
+
+			HedgedRequestsExchangeFilterFunction filter = new HedgedRequestsExchangeFilterFunction(hedgingPolicy, Collections
+					.singletonList(metricsReporter));
+			return filter.filter(clientRequest, exchangeFunction);
+		})
+				.expectSubscription()
+				.expectNoEvent(Duration.of(200, ChronoUnit.MILLIS))
+				.expectNext(response)
+				.verifyComplete();
+
+		verify(hedgingPolicy).record(clientRequest, response, 200, null);
+		verify(metricsReporter).record(clientRequest, response, 200, null);
+	}
+
+	@Test
+	public void hedgeOnceOriginalNeverReturns() {
+		when(hedgingPolicy.shouldHedge(clientRequest)).thenReturn(true);
+		when(hedgingPolicy.getNumberOfHedgedRequests(clientRequest)).thenReturn(1);
+		when(hedgingPolicy.getDelayBeforeHedging(clientRequest)).thenReturn(Duration.ofMillis(100));
+
+		ClientResponse response = mock(ClientResponse.class);
+		StepVerifier.withVirtualTime(() -> {
+
+			AtomicInteger count = new AtomicInteger(0);
+			Mono<ClientResponse> first = Mono.never();
+			Mono<ClientResponse> second = Mono.just(response).delayElement(Duration.ofMillis(150));
+			when(exchangeFunction.exchange(clientRequest)).then(invocation -> count.getAndIncrement() == 0 ? first : second);
+
+			HedgedRequestsExchangeFilterFunction filter = new HedgedRequestsExchangeFilterFunction(hedgingPolicy, Collections
+					.singletonList(metricsReporter));
+			return filter.filter(clientRequest, exchangeFunction);
+		})
+				.expectSubscription()
+				.expectNoEvent(Duration.ofMillis(250))
+				.expectNext(response)
+				.verifyComplete();
+
+		verify(hedgingPolicy).record(clientRequest, response, 150, 1);
+		verify(metricsReporter).record(clientRequest, response, 150, 1);
+	}
+
+	@Test
+	public void firstRequestSucceedsNeverHedge() {
+		when(hedgingPolicy.shouldHedge(clientRequest)).thenReturn(true);
+		when(hedgingPolicy.getNumberOfHedgedRequests(clientRequest)).thenReturn(2);
+		when(hedgingPolicy.getDelayBeforeHedging(clientRequest)).thenReturn(Duration.ofMillis(100));
+
+		ClientResponse response = mock(ClientResponse.class);
+		StepVerifier.withVirtualTime(() -> {
+			when(exchangeFunction.exchange(clientRequest)).thenReturn(Mono.just(response).delayElement(Duration.ofMillis(50)));
+
+			HedgedRequestsExchangeFilterFunction filter = new HedgedRequestsExchangeFilterFunction(hedgingPolicy, Collections
+					.singletonList(metricsReporter));
+			return filter.filter(clientRequest, exchangeFunction);
+		})
+				.expectSubscription()
+				.expectNoEvent(Duration.ofMillis(50))
+				.expectNext(response)
+				.verifyComplete();
+
+		verify(hedgingPolicy).record(clientRequest, response, 50, null);
+		verify(metricsReporter).record(clientRequest, response, 50, null);
+		verify(exchangeFunction, times(1)).exchange(clientRequest);
+	}
+
```

Comment:
What about the case where multiple (or all succeed)?

Reply:
What would we asset in that case?